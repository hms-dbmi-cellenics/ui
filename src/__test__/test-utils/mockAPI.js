import _ from 'lodash';

// Default platform data
import cellSetsData from '__test__/data/cell_sets.json';
import backendStatusData from '__test__/data/backend_status.json';
import processingConfigData from '__test__/data/processing_config.json';
import mockDemoExperiments from '__test__/test-utils/mockData/mockDemoExperiments.json';

import fake from '__test__/test-utils/constants';

import {
  responseData,
} from '__test__/test-utils/mockData';

const promiseResponse = (
  response,
  options = {},
) => Promise.resolve(new Response(response, options));

const statusResponse = (code, body) => (
  Promise.resolve({
    status: code,
    body: JSON.stringify(body),
  })
);

const delayedResponse = (response, delay = 10000, options = {}) => new Promise((resolve) => {
  setTimeout(() => {
    resolve(new Response(response, options));
  }, delay);
});

const workerDataResult = (data) => Promise.resolve(_.cloneDeep(data));

const fetchWorkMock = (
  mockedResults,
) => ((experimentId, body) => {
  if (typeof mockedResults[body.name] === 'function') {
    return mockedResults[body.name]();
  }

  return workerDataResult(mockedResults[experimentId]);
});

const generateDefaultMockAPIResponses = (experimentId) => ({
  [`experiments/${experimentId}$`]: () => promiseResponse(
    JSON.stringify(responseData.experiments.find(({ id }) => id === experimentId)),
  ),
  [`experiments/${experimentId}/processingConfig$`]: () => promiseResponse(
    JSON.stringify(processingConfigData),
  ),
  [`experiments/${experimentId}/cellSets$`]: () => promiseResponse(
    JSON.stringify(cellSetsData),
  ),
  [`experiments/${experimentId}/backendStatus$`]: () => promiseResponse(
    JSON.stringify(backendStatusData),
  ),
  experiments$: () => promiseResponse(
    JSON.stringify(responseData.experiments),
  ),
  [`experiments/${experimentId}/samples$`]: () => promiseResponse(
    JSON.stringify(responseData.samples[0]),
  ),
  '/v2/experiments/examples$': () => promiseResponse(
    JSON.stringify(mockDemoExperiments),
  ),
  'experiments/clone$': () => promiseResponse(
    JSON.stringify(fake.CLONED_EXPERIMENT_ID),
  ),
  [`/v2/workRequest/${experimentId}`]: () => statusResponse(200, 'OK'),
});

const mockAPI = (apiMapping) => (req) => {
  const path = req.url;

  const key = _.find(Object.keys(apiMapping), (matcherStr) => new RegExp(matcherStr).test(path));

  if (!key) {
    return statusResponse({
      status: 404,
      body: `Path ${path} is undefined`,
    });
  }

  return apiMapping[key](req);
};

export default mockAPI;
export {
  generateDefaultMockAPIResponses,
  promiseResponse,
  statusResponse,
  delayedResponse,
  workerDataResult,
  fetchWorkMock,
};
