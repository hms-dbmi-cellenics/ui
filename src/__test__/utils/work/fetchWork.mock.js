const mockGenesListData = ['A', 'B', 'C', 'D'];

// const mockGeneExpressionData = {
//   A: {
//     rawExpression: {
//       min: 0,
//       max: 6.8,
//       mean: 1.68,
//       stdev: 2.597331964,
//       expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
//     },
//     truncatedExpression: {
//       min: 0,
//       max: 6.8,
//       mean: 1.68,
//       stdev: 2.597331964,
//       expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
//     },
//   },
//   B: {
//     rawExpression: {
//       min: 0,
//       max: 6.8,
//       mean: 1.702857143,
//       stdev: 2.551115536,
//       expression: [0, 0, 0, 2.56, 0, 6.8, 2.56],
//     },
//     truncatedExpression: {
//       min: 0,
//       max: 6.8,
//       mean: 1.702857143,
//       stdev: 2.551115536,
//       expression: [0, 0, 0, 2.56, 0, 6.8, 2.56],
//     },
//   },
//   C: {
//     rawExpression: {
//       min: 0,
//       max: 3.4,
//       mean: 1.68,
//       stdev: 2.141525936,
//       expression: [0, 0, 0, 3.56, 0, 4.8, 3.4],
//     },
//     truncatedExpression: {
//       min: 0,
//       max: 3.4,
//       mean: 1.68,
//       stdev: 2.141525936,
//       expression: [0, 0, 0, 3.56, 0, 4.8, 3.4],
//     },
//   },
//   D: {
//     rawExpression: {
//       min: 0,
//       max: 6.8,
//       mean: 1.68,
//       stdev: 2.597331964,
//       expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
//     },
//     truncatedExpression: {
//       min: 0,
//       max: 6.8,
//       mean: 1.68,
//       stdev: 2.597331964,
//       expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
//     },
//   },
//   E: { hello: 'world' },
// };

// const mockCacheGet = jest.fn((x) => {
//   const mockCacheContents = {
//     fd3161a878f67ebf54018720cffd6a66: 'A', // pragma: allowlist secret
//     '10250a11679234110a1c260d6fd81d3c': 'B', // pragma: allowlist secret
//     f5c957411a28de68f35e1f5c8a29da7e: 'C', // pragma: allowlist secret
//     '33faa711a94a2028b5bae1778126aec0': 'E', // pragma: allowlist secret
//   };

//   if (x in mockCacheContents) {
//     return mockGeneExpressionData[mockCacheContents[x]];
//   }

//   return null;
// });

const mockCacheGet = jest.fn();
const mockCacheSet = jest.fn();
const mockCacheRemove = jest.fn();

const mockFetchWorkRequest = jest.fn(() => true);

const mockSeekFromS3 = jest.fn();

const mockCacheModule = {
  get: jest.fn((x) => mockCacheGet(x)),
  set: jest.fn((key, val) => mockCacheSet(key, val)),
  _remove: jest.fn((key) => mockCacheRemove(key)),
};

const mockFetchWorkModule = {
  __esModule: true,
  fetchWork: mockFetchWorkRequest,
};

const mockQcPipelineStartDate = '2021-01-01T01:01:01.000Z';

export {
  mockGenesListData, mockCacheGet, mockCacheSet,
  mockFetchWorkRequest, mockSeekFromS3,
  mockQcPipelineStartDate, mockCacheModule, mockFetchWorkModule,
};
