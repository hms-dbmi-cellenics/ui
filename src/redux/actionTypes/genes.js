const GENES = 'genes';

/**
 * Turns on the loading condition for gene information.
 * Components that have the particular properties requested in their dependencies
 * should set a loading state at this event.
 */
const GENES_PROPERTIES_LOADING = `${GENES}/propertiesLoading`;

/**
 * Sets the state of the gene data to be successfully loaded, with the appropriate gene data.
 * This is meant to be used for paginated data, where the data is
 * lazy-loaded by an API call or equivalent.
 */
const GENES_PROPERTIES_LOADED_PAGINATED = `${GENES}/propertiesLoadedPaginated`;

/**
 * Sets an error condition for gene properties data.
 */
const GENES_PROPERTIES_ERROR = `${GENES}/propertiesError`;

/**
 * Updates the list of selected genes with a new selected gene.
 */
const GENES_SELECT = `${GENES}/select`;

/**
 * Removes a deselected gene from the list of selected genes.
 */
const GENES_DESELECT = `${GENES}/deselect`;

/**
 * Turns on the loading state for expression data.
 */
const GENES_EXPRESSION_LOADING = `${GENES}/expressionLoading`;

/**
 * Sets the state of the expression store to be successfully loaded,
 * with the appropriate expression data.
 */
const GENES_EXPRESSION_LOADED = `${GENES}/expressionLoaded`;

/**
 * Sets an error condition for gene expression data.
 */
const GENES_EXPRESSION_ERROR = `${GENES}/expressionError`;

/**
 * Sets marker genes as loading
 */
const MARKER_GENES_LOADING = `${GENES}/markerGenesLoading`;

/**
 * Marker genes have finished loading
 */
const MARKER_GENES_LOADED = `${GENES}/markerGenesLoaded`;

/**
 * Marker genes load task had some kind of error
 */
const MARKER_GENES_ERROR = `${GENES}/markerGenesError`;

/**
 * Downsampled genes expression load task
 */
const DOWNSAMPLED_GENES_EXPRESSION_LOADING = `${GENES}/downsampledGenesExpressionLoading`;

/**
 * Downsampled genes expression loaded
 */
const DOWNSAMPLED_GENES_EXPRESSION_LOADED = `${GENES}/downsampledGenesExpressionLoaded`;

/**
 * Downsampled genes expression loaded
 */
const DOWNSAMPLED_GENES_EXPRESSION_ERROR = `${GENES}/downsampledGenesExpressionError`;

export {
  GENES_PROPERTIES_LOADING, GENES_PROPERTIES_LOADED_PAGINATED, GENES_PROPERTIES_ERROR,
  GENES_SELECT, GENES_DESELECT,
  MARKER_GENES_LOADING, MARKER_GENES_LOADED, MARKER_GENES_ERROR,
  GENES_EXPRESSION_LOADING, GENES_EXPRESSION_LOADED, GENES_EXPRESSION_ERROR,
  DOWNSAMPLED_GENES_EXPRESSION_LOADING,
  DOWNSAMPLED_GENES_EXPRESSION_LOADED,
  DOWNSAMPLED_GENES_EXPRESSION_ERROR,
};
