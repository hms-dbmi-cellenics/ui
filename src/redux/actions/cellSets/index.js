import createCellSet from './createCellSet';
import deleteCellSet from './deleteCellSet';
import deleteCellClass from './deleteCellClass';
import runCellSetsClustering from './runCellSetsClustering';
import runCellSetsAnnotation from './runCellSetsAnnotation';
import runCellCycleScoring from './runCellCycleScoring';
import updateCellSetsClustering from './updateCellSetsClustering';

import loadCellSets from './loadCellSets';
import reorderCellSet from './reorderCellSet';
import updateCellSetProperty from './updateCellSetProperty';
import updateCellSetSelected from './updateCellSetSelected';
import setCellSetHiddenStatus from './setCellSetHiddenStatus';
import unhideAllCellSets from './unhideAllCellSets';

export {
  createCellSet,
  deleteCellSet,
  deleteCellClass,
  runCellSetsClustering,
  runCellSetsAnnotation,
  updateCellSetsClustering,
  loadCellSets,
  reorderCellSet,
  updateCellSetSelected,
  updateCellSetProperty,
  setCellSetHiddenStatus,
  unhideAllCellSets,
  runCellCycleScoring,
};
