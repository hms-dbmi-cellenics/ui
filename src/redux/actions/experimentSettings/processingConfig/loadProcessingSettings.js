import handleError from 'utils/http/handleError';
import fetchAPI from 'utils/http/fetchAPI';

import endUserMessages from 'utils/endUserMessages';
import {
  EXPERIMENT_SETTINGS_PROCESSING_CONFIG_LOADED,
  EXPERIMENT_SETTINGS_PROCESSING_ERROR,
} from 'redux/actionTypes/experimentSettings';

const loadProcessingSettings = (experimentId) => async (dispatch) => {
  const url = `/v1/experiments/${experimentId}/processingConfig`;

  try {
    const data = await fetchAPI(url);

    dispatch({
      type: EXPERIMENT_SETTINGS_PROCESSING_CONFIG_LOADED,
      payload: {
        data: data.processingConfig,
      },
    });

    return;
  } catch (e) {
    const errorMessage = handleError(e, endUserMessages.ERROR_FETCHING_DATA_PROCESSING_SETTINGS);

    dispatch({
      type: EXPERIMENT_SETTINGS_PROCESSING_ERROR,
      payload: {
        error: errorMessage,
      },
    });
  }
};

export default loadProcessingSettings;
