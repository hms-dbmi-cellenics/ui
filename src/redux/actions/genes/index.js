import loadGeneExpression from './loadGeneExpression';
import changeGeneSelection from './changeGeneSelection';
import loadPaginatedGeneProperties from './loadPaginatedGeneProperties';
import loadMarkerGenes from './loadMarkerGenes';
import loadDownsampledGeneExpression from './loadDownsampledGeneExpression';

export {
  loadGeneExpression,
  changeGeneSelection,
  loadPaginatedGeneProperties,
  loadMarkerGenes,
  loadDownsampledGeneExpression,
};
