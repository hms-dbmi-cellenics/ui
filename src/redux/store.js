/* eslint-disable global-require */
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { createWrapper } from 'next-redux-wrapper';
import { enableMapSet } from 'immer';

import rootReducer from 'redux/reducers/index';

enableMapSet();

const bindMiddleware = (middleware) => {
  const { composeWithDevTools } = require('redux-devtools-extension');

  // eslint-disable-next-line import/no-extraneous-dependencies
  const { createLogger } = require('redux-logger');

  // do not log server-side redux actions
  middleware.push(createLogger({
    predicate: () => typeof window !== 'undefined',
  }));

  return composeWithDevTools(applyMiddleware(...middleware));
};

const makeStore = () => {
  const store = createStore(
    rootReducer,
    bindMiddleware([thunk]),
  );

  // IF REDUCERS WERE CHANGED, RELOAD WITH INITIAL STATE
  if (module.hot) {
    module.hot.accept('./reducers', () => {
      const createNextReducer = require('./reducers').default;
      store.replaceReducer(createNextReducer(rootReducer));
    });
  }
  return store;
};

const wrapper = createWrapper(makeStore, { debug: false });
export { wrapper, makeStore };
