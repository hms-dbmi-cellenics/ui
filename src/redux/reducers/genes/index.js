import getInitialState from 'redux/reducers/genes/getInitialState';
import {
  GENES_PROPERTIES_LOADING, GENES_PROPERTIES_LOADED_PAGINATED, GENES_PROPERTIES_ERROR,
  GENES_SELECT, GENES_DESELECT,
  GENES_EXPRESSION_LOADING, GENES_EXPRESSION_LOADED, GENES_EXPRESSION_ERROR,
  MARKER_GENES_LOADING, MARKER_GENES_LOADED, MARKER_GENES_ERROR,
  DOWNSAMPLED_GENES_EXPRESSION_LOADING,
  DOWNSAMPLED_GENES_EXPRESSION_LOADED,
  DOWNSAMPLED_GENES_EXPRESSION_ERROR,
} from 'redux/actionTypes/genes';

import { EXPERIMENT_SETTINGS_QC_START } from 'redux/actionTypes/experimentSettings';

import genesPropertiesLoading from 'redux/reducers/genes/genesPropertiesLoading';
import genesPropertiesLoadedPaginated from 'redux/reducers/genes/genesPropertiesLoadedPaginated';
import genesPropertiesError from 'redux/reducers/genes/genesPropertiesError';

import genesExpressionLoading from 'redux/reducers/genes/genesExpressionLoading';
import genesExpressionLoaded from 'redux/reducers/genes/genesExpressionLoaded';
import genesExpressionError from 'redux/reducers/genes/genesExpressionError';

import downsampledGenesLoading from 'redux/reducers/genes/downsampledGenesLoading';
import downsampledGenesLoaded from 'redux/reducers/genes/downsampledGenesLoaded';

import markerGenesLoading from 'redux/reducers/genes/markerGenesLoading';
import markerGenesError from 'redux/reducers/genes/markerGenesError';
import markerGenesLoaded from 'redux/reducers/genes/markerGenesLoaded';

import genesSelect from 'redux/reducers/genes/genesSelect';
import genesDeselect from 'redux/reducers/genes/genesDeselect';
import downsampledGenesError from './downsampledGenesError';

const genesReducer = (state = getInitialState(), action) => {
  switch (action.type) {
    case GENES_PROPERTIES_LOADING: {
      return genesPropertiesLoading(state, action);
    }
    case GENES_PROPERTIES_LOADED_PAGINATED: {
      return genesPropertiesLoadedPaginated(state, action);
    }
    case GENES_PROPERTIES_ERROR: {
      return genesPropertiesError(state, action);
    }
    case GENES_SELECT: {
      return genesSelect(state, action);
    }
    case GENES_DESELECT: {
      return genesDeselect(state, action);
    }
    case GENES_EXPRESSION_LOADING: {
      return genesExpressionLoading(state, action);
    }
    case GENES_EXPRESSION_LOADED: {
      return genesExpressionLoaded(state, action);
    }
    case GENES_EXPRESSION_ERROR: {
      return genesExpressionError(state, action);
    }
    case EXPERIMENT_SETTINGS_QC_START: {
      return getInitialState();
    }
    case MARKER_GENES_LOADING: {
      return markerGenesLoading(state, action);
    }
    case DOWNSAMPLED_GENES_EXPRESSION_LOADING: {
      return downsampledGenesLoading(state, action);
    }
    case DOWNSAMPLED_GENES_EXPRESSION_LOADED: {
      return downsampledGenesLoaded(state, action);
    }
    case DOWNSAMPLED_GENES_EXPRESSION_ERROR: {
      return downsampledGenesError(state, action);
    }
    case MARKER_GENES_LOADED: {
      return markerGenesLoaded(state, action);
    }
    case MARKER_GENES_ERROR: {
      return markerGenesError(state, action);
    }
    default: {
      return state;
    }
  }
};

export default genesReducer;
