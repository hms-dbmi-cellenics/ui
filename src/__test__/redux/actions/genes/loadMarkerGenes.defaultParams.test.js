import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';
import mockAPI, {
  generateDefaultMockAPIResponses,
  promiseResponse,
  workerDataResult,
} from '__test__/test-utils/mockAPI';

import { MD5 } from 'object-hash';
import _ from 'lodash';
import backendStatusData from '__test__/data/backend_status.json';
import createObjectHash from 'utils/work/createObjectHash';
import generateExperimentSettingsMock from '__test__/test-utils/experimentSettings.mock';
import { loadBackendStatus } from 'redux/actions/backendStatus';
import { loadMarkerGenes } from 'redux/actions/genes';
import loadProcessingSettings from 'redux/actions/experimentSettings/processingConfig/loadProcessingSettings';
import { makeStore } from 'redux/store';
import { dispatchWorkRequest } from 'utils/work/seekWorkResponse';
import { loadCellSets } from 'redux/actions/cellSets';

jest.mock('utils/getTimeoutForWorkerTask', () => ({
  __esModule: true, // this property makes it work
  default: () => 60,
}));

enableFetchMocks();

jest.mock('utils/work/seekWorkResponse', () => ({
  __esModule: true, // this property makes it work
  dispatchWorkRequest: jest.fn(),
}));

jest.mock('utils/work/createObjectHash');

const experimentId = '6463cb35-3e08-4e94-a181-6d155a5ca570';
const sampleIds = ['sample-WT', 'sample-WT1', 'sample-KO'];
const initialExperimentSettingsState = generateExperimentSettingsMock(sampleIds);

const experimentSettings = {
  ...initialExperimentSettingsState,
  processing: {
    ...initialExperimentSettingsState.processing,
    configureEmbedding: {
      clusteringSettings: {
        method: 'louvain',
        methodSettings: {
          louvain: {
            resolution: 0.8,
          },
        },
      },
      embeddingSettings: {
        method: 'umap',
        methodSettings: {
          tsne: { perplexity: 30, learningRate: 200 },
          umap: { distanceMetric: 'cosine', minimumDistance: 0.3 },
        },
      },
    },
  },
};

// this is the same Date used in the API to make sure the default ETag generated
// by the UI is identical to the API
const date = new Date(1458619200000);
backendStatusData.pipeline.startDate = date.toISOString();

const customAPIResponses = {
  [`experiments/${experimentId}/processingConfig$`]: () => promiseResponse(JSON.stringify(experimentSettings.processing)),
  [`experiments/${experimentId}/backendStatus$`]: () => promiseResponse(
    JSON.stringify(backendStatusData),
  ),
};

const mockApiResponses = _.merge(
  generateDefaultMockAPIResponses(experimentId), customAPIResponses,
);

describe('loadEmbedding action', () => {
  let store;

  beforeEach(async () => {
    jest.clearAllMocks();
    fetchMock.resetMocks();
    fetchMock.doMock();
    fetchMock.mockIf(/.*/, mockAPI(mockApiResponses));

    dispatchWorkRequest
      .mockReset()
      .mockImplementationOnce(() => workerDataResult([[1, 2], [3, 4]]));
    createObjectHash.mockImplementation((object) => MD5(object));

    store = makeStore();
    await store.dispatch(loadProcessingSettings(experimentId));
    await store.dispatch(loadBackendStatus(experimentId));
    await store.dispatch(loadCellSets(experimentId));
  });

  it('loadEmbedding generates correct hash params body for ETag', async () => {
    const hashMock = jest.fn((object) => MD5(object));
    createObjectHash.mockImplementation(hashMock);

    await store.dispatch(loadMarkerGenes(experimentId, 'interactiveHeatmap'));

    expect(hashMock).toHaveBeenCalled();
    // this snapshot should max exactly API snapshot:
    // submitMarkerHeatmapWork.test.js.snap
    expect(hashMock.mock.calls).toMatchSnapshot('marker heatmap - hash call params');
    // this ETag should match exactly the one in
    // submitMarkerHeatmap.test.js
    const ETag = hashMock.mock.results[0].value;
    expect(ETag).toEqual('5a0e619279cd36c96c0ffd89d0d86d46'); // pragma: allowlist secret`
  });
});
