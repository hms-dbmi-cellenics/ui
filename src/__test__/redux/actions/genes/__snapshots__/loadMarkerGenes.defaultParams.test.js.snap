// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadEmbedding action loadEmbedding generates correct hash params body for ETag: marker heatmap - hash call params 1`] = `
[
  [
    {
      "body": {
        "downsampleSettings": {
          "groupedTracks": [
            "sample",
            "louvain",
          ],
          "hiddenCellSets": [],
          "selectedCellSet": "louvain",
          "selectedPoints": "All",
        },
        "nGenes": 5,
        "name": "MarkerHeatmap",
      },
      "cacheUniquenessKey": null,
      "experimentId": "6463cb35-3e08-4e94-a181-6d155a5ca570",
      "extraDependencies": [
        {
          "method": "louvain",
          "methodSettings": {
            "louvain": {
              "resolution": 0.8,
            },
          },
        },
        [
          [
            "louvain-0",
            "louvain-1",
            "louvain-2",
            "louvain-3",
            "louvain-4",
            "louvain-5",
            "louvain-6",
            "louvain-7",
            "louvain-8",
            "louvain-9",
            "louvain-10",
            "louvain-11",
            "louvain-12",
            "louvain-13",
          ],
          [
            "louvain-0",
            "louvain-1",
            "louvain-2",
            "louvain-3",
            "louvain-4",
            "louvain-5",
            "louvain-6",
            "louvain-7",
            "louvain-8",
            "louvain-9",
            "louvain-10",
            "louvain-11",
            "louvain-12",
            "louvain-13",
            "b62028a1-ffa0-4f10-823d-93c9ddb88898",
            "ab568662-27fa-462c-9435-625594341314",
            "98c7a0ed-d086-4df8-bf94-a9ee6edb793f",
            "Track_1-KMeta",
            "Track_1-WMetaT",
          ],
        ],
        1,
      ],
      "extras": undefined,
      "qcPipelineStartDate": "2016-03-22T04:00:00.000Z",
      "workerVersion": 4,
    },
  ],
]
`;
